src/rdo.rs

```diff
@@ -355,9 +355,48 @@ fn compute_tx_distortion<T: Pixel>(
   distortion
 }
 
-pub fn compute_rd_cost<T: Pixel>(fi: &FrameInvariants<T>, rate: u32, distortion: u64) -> f64 {
+pub fn compute_rd_cost_biased_by_importance<T: Pixel>(
+  fi: &FrameInvariants<T>, rect: Rect, rate: u32, distortion: u64
+) -> f64 {
+  assert!(rect.x >= 0);
+  assert!(rect.y >= 0);
+  assert!(rect.x % 4 == 0);
+  assert!(rect.y % 4 == 0);
+  assert!(rect.width % 4 == 0);
+  assert!(rect.height % 4 == 0);
+
+  // eprintln!("{:?}", rect);
+  // eprintln!("{}×{}", fi.w_in_b, fi.h_in_b);
+  let x1 = rect.x as usize / 4;
+  let y1 = rect.y as usize / 4;
+  let x2 = (x1 + rect.height / 4).min(fi.w_in_b);
+  let y2 = (y1 + rect.height / 4).min(fi.h_in_b);
+
+  let mut mean_importance = 0.;
+  let mut count = 0;
+  for y in y1..y2 {
+    for x in x1..x2 {
+      mean_importance += fi.block_importances[y * fi.w_in_b + x];
+      count += 1;
+    }
+  }
+  mean_importance /= count as f32;
```

Comment:
Divides are expensive. Take advantage of the fact that block areas are always a power of 2. (see BlockSize::width_log2 and height_log2)

Reply:
It would be easier if I didn't have to `min()` the end coordinates against `w_in_b` and `h_in_b`, respectively. Not doing that leads to asserts, for instance in the flushing tests.

Reply:
Ah right, I think although partitioning is forced on the edges, it's allowed to be split vertically / horizontally, which can allow quite a few blocks out of the frame. One option here is to add additional padding to your importance structure so you can drop the checks.

Reply:
Also here I think you want "count" to cover the whole area that was requested, not just the available area.

Reply:
So the out-of-bounds blocks would count as zero importance?

Reply:
Or whatever your "default" importance is (a block not referenced by any other frame)

Reply:
Right now I compute importance in `f32`s, I think I would need to compute it in `u64`s or `u32`s like x264 does to replace this with a shift.

Reply:
It would be optimal.

Reply:
OK. Let's wait for a future patch to replace it with a shift. Then we can show there's no quality loss for doing so.